File: app/src/main/java/com/xattacker/util/WeakReferenceList.java
package com.xattacker.util;

import java.lang.ref.WeakReference;
import java.util.ArrayList;

public final class WeakReferenceList <T>
{
	public interface WeakReferenceListVisitor <T>
	{
		void onReferenceFetched(T aReference);
	}
	
	
	private ArrayList<WeakReference<T>> _references;
	
	public WeakReferenceList()
	{
		_references = new ArrayList<WeakReference<T>>();
	}
	
	public boolean isEmpty()
	{
		return _references == null || _references.isEmpty();
	}

	public int size()
	{
		return _references != null ? _references.size() : 0;
	}

	public void clearReferences()
	{
		if (_references != null)
		{
			_references.clear();
		}
	}

	public T get(int aIndex) throws NullPointerException
	{
		return _references != null && _references.size() > aIndex ? _references.get(aIndex).get() : null;
	}

	public void addReference(T aReference)
	{
		if (aReference != null && _references != null)
		{
			WeakReference<T> listener = null;
			boolean existed = false;
			
			for (int i = 0; i < _references.size(); i++)
			{
				listener = _references.get(i);
				
				if (listener != null && listener.get() != null)
				{
					if (listener.get() == aReference)
					{
						existed = true;
					
						break;
					}
				}
				else
				{
					_references.remove(i);
					i--;
				}
			}
			
			if (!existed)
			{		
				_references.add(new WeakReference<T>(aReference));
			}
		}
	}
	
	public void removeReference(T aReference)
	{
		if (
			aReference != null &&
			_references != null &&
			!_references.isEmpty()
			)
		{
			WeakReference<T> listener = null;
			
			for (int i = 0; i < _references.size(); i++)
			{
				listener = _references.get(i);
				
				if (listener != null && listener.get() != null)
				{
					if (listener.get() == aReference)
					{
						_references.remove(listener);
					
						break;
					}
				}
				else
				{
					_references.remove(i);
					i--;
				}
			}
		}
	}
	
	public void fetchReference(WeakReferenceListVisitor<T> aVisitor)
	{
		if (aVisitor != null && _references != null && !_references.isEmpty())
		{
			 WeakReference<T> reference = null;
				
		     for (int i = 0; i < _references.size(); i++)
			 {
		   	 	 reference = _references.get(i);
				
				 if (reference != null && reference.get() != null)
				 {
					 aVisitor.onReferenceFetched(reference.get());
				 }
				 else
				 {
					 _references.remove(i);
					 i--;
				 }
			 }
		}
	}
}

File: app/src/main/java/com/xattacker/util/WeakReferenceList.kt
package com.xattacker.util

import java.lang.ref.WeakReference
import java.util.*

class WeakReferenceList<T>
{
    private val _references: ArrayList<WeakReference<T>>?

    val isEmpty: Boolean get() = _references == null || _references.isEmpty()

    init
    {
        _references = ArrayList()
    }

    fun count(): Int
    {
        return _references?.size ?: 0
    }

    fun get(aIndex: Int): T?
    {
        return _references?.get(aIndex)?.get()
    }

    fun clear()
    {
        _references?.clear()
    }

    fun addReference(aReference: T?)
    {
        if (aReference != null && _references != null)
        {
            var listener: WeakReference<T>?
            var existed = false
            var i = 0

            while (i < _references.size)
            {
                listener = _references[i]

                if (listener.get() != null)
                {
                    if (listener.get() === aReference)
                    {
                        existed = true

                        break
                    }
                }
                else
                {
                    _references.removeAt(i)
                    i--
                }

                i++
            }

            if (!existed)
            {
                _references.add(WeakReference(aReference))
            }
        }
    }

    fun removeReference(aReference: T?)
    {
        if (aReference != null && _references != null && !_references.isEmpty())
        {
            var listener: WeakReference<T>?
            var i = 0

            while (i < _references.size)
            {
                listener = _references[i]

                if (listener.get() != null)
                {
                    if (listener.get() === aReference)
                    {
                        _references.remove(listener)

                        break
                    }
                }
                else
                {
                    _references.removeAt(i)
                    i--
                }

                i++
            }
        }
    }

    fun fetch(aVisitor: (T) -> Unit)
    {
        if (_references != null && !_references.isEmpty())
        {
            var i = 0

            while (i < _references.size)
            {
                val ref = _references[i].get()

                if (ref != null)
                {
                    aVisitor.invoke(ref)
                }
                else
                {
                    _references.removeAt(i)
                    i--
                }

                i++
            }
        }
    }
}
